<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2024-03-30 Sat 03:25 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>&lrm;</title>
<meta name="author" content="Pjotr Prins" />
<meta name="generator" content="Org Mode" />
<style>
  #content { max-width: 60em; margin: auto; }
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #e6e6e6;
    border-radius: 3px;
    background-color: #f2f2f2;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: auto;
  }
  pre.src:before {
    display: none;
    position: absolute;
    top: -8px;
    right: 12px;
    padding: 3px;
    color: #555;
    background-color: #f2f2f299;
  }
  pre.src:hover:before { display: inline; margin-top: 14px;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-authinfo::before { content: 'Authinfo'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .equation-container {
    display: table;
    text-align: center;
    width: 100%;
  }
  .equation {
    vertical-align: middle;
  }
  .equation-label {
    display: table-cell;
    text-align: right;
    vertical-align: middle;
  }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { }
</style>
</head>
<body>
<div id="content" class="content">
<div id="table-of-contents" role="doc-toc">
<h2>Table of Contents</h2>
<div id="text-table-of-contents" role="doc-toc">
<ul>
<li><a href="#orge5fc3fa">1. HEGP challenge</a></li>
<li><a href="#org5bb9b2c">2. Details</a></li>
<li><a href="#orgd1fd6d2">3. Reference code</a></li>
<li><a href="#orgfa78c7c">4. Submissions</a></li>
<li><a href="#orga65d5e4">5. Feedback</a></li>
<li><a href="#org24f8220">6. General introduction</a></li>
<li><a href="#orgf59437b">7. Toy Example</a></li>
<li><a href="#org2c4b78f">8. References</a></li>
</ul>
</div>
</div>
<div id="outline-container-orge5fc3fa" class="outline-2">
<h2 id="orge5fc3fa"><span class="section-number-2">1.</span> HEGP challenge</h2>
<div class="outline-text-2" id="text-1">
<p>
<i>2020-11-01</i> The authors announce the HEGP (heegeepee;
<i>hiːdʒiːpiː</i>) Challenge, with a $1,000 (one thousand dollar) prize for
the individual or group who can crack HEGP encrypted data.
</p>

<p>
<i>See this <a href="./start">page</a> for a general introduction and relevance of HEGP</i>
</p>
</div>
</div>

<div id="outline-container-org5bb9b2c" class="outline-2">
<h2 id="org5bb9b2c"><span class="section-number-2">2.</span> Details</h2>
<div class="outline-text-2" id="text-2">
<p>
The authors offer a reward of $1000 (one thousand US dollar) for
cracking the HEGP homomorphic encryption method for genotypes and
phenotypes as described in this open access <a href="https://www.genetics.org/content/215/2/359">paper</a>.
</p>

<p>
The challenge consists of decryption of two genotype data sets.
</p>

<p>
In both data sets, plain text genotypes are represented numerically as
matrices of allele dosages, with rows representing individuals and
columns representing <a href="https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism">SNP sites</a> or <a href="https://en.wikipedia.org/wiki/Genotype">genotype</a> markers (see below toy
example). Each plain text genotype dosage takes the integer values
0,1,2. After encryption these dosages are transformed into floating
point cyphertexts that are approximately normally distributed. The
challenge is to recover the unencrypted original values from the
cyphertext.
</p>

<p>
In HEGP, the ciphertext has been generated by matrix multiplication of
the plaintext genotype matrix by a random orthogonal matrix (the key
generated using Linux <code>/dev/random</code> and <a href="https://github.com/encryption4genetics/hegp-R/blob/9989c92f1c844fd5d8f6724be1496b1aba07b125/hegp/R/hegp.R#L168">rustiefel</a>) which is unknown to
you. The task therefore is to find the inverse of this matrix, which
because it is orthogonal, will be the transpose of original key.
</p>

<p>
For the first challenge1 we encrypted a data set consisting of 500
individuals by 12,359 SNP genotypes which exists in the public
domain. We consider this data cracked if 50 individuals (10%) are
identified correctly. Note that this data might be from human,
mouse, rat, nematode or plants. Note also that these genotypes
are a subset of a public dataset, and that any missing genotypes in
the public data have been imputed before encryption.
</p>

<p>
The second challenge2 may be harder. We encrypted a mammalian data set
that is not in the public domain (yet), consisting of 1,586
individuals and 12,561 SNPs. We consider the code cracked if you can
compute the plaintext genotypes of this matrix correctly with mean
error of under 10%.
</p>

<p>
The ciphertext files for challenge1 and challenge2 can be downloaded from
<a href="https://files.genenetwork.org/HEGP-challenge/">https://files.genenetwork.org/HEGP-challenge/</a>
</p>

<p>
The MD5 checksums of the ciphertext files are found in md5sum.txt. The
full list, also for the original (secret for now) plaintext and keys
is
</p>

<pre class="example" id="org67afc48">
5605626a9403ebdf96041d2dd766a516  challenge1.ciphertext.RData
148ce09c6dfc878af09874c090a7bd64  challenge1.ciphertext.txt
abbc5ce33025e723805cb14a332f6c9d  challenge1.maf.txt
29f348fd4d9fbd453f658def53dbb87a  challenge1.key.RData
b03e91f69d8124658e30562e1fa3b64c  challenge1.key.txt
dcf724ff4798f4ec6674d029f220385f  challenge1.plaintext.RData
e26367b991dfcedeaee139ffcc9a38a7  challenge1.plaintext.txt
d2db18e96a3a8173b12383aa8d606583  challenge1.sample.ids.txt

6d5d10279424a7c952ed8f9b0b086e84  challenge2.ciphertext.RData
c3706ecf12275859e3a040b23af1c9f9  challenge2.ciphertext.txt
15fa26a4d03e18f05643e917b8480216  challenge2.maf.txt
9615196bf89fb19105686a018e4c40c0  challenge2.key.RData
ecf00d15e233b6dda73e40ebdc3358d2  challenge2.key.txt
a007e8f5f085798a88180386ffbfa58f  challenge2.plaintext.RData
2b1870215dd67cfe1b19c423979ab3c7  challenge2.plaintext.txt
</pre>

<p>
With this challenge we want to address the concern that individuals
might be identified based on information in the public domain and/or
by finding patterns in the encrypted data.
</p>

<p>
We will release the plaintext file and the key file when the challenge
is cracked or after six months (which is May 1st 2021). In case the prize
is unclaimed we'll create a new dataset for a fresh challenge.
</p>

<p>
The ciphertext Datasets are provided as RData objects which can be
manipulated using the hegp R package. Briefly, a dataset D is a list,
comprising of minimally a phenotype vector D$y, and a genotype dosage
matrix D$geno (rows are individuals, columns are SNPs). There are
additional components that are useful for GWAS but not relevant for
the decryption challenge. Full details are in the documentation of the
R package. The challenge datasets have fake phenotype data that can
be ignored.
</p>

<p>
To unpack the RData original datasets and write the genotypes and
allele frequencies to tab-delimited text files from R, convert with
</p>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #8bc34a;">library</span>(readr)
<span style="color: #8bc34a;">load</span>(<span style="color: #9ccc65;">"challenge2.ciphertext.RData"</span>)
write.csv( De$geno, <span style="color: #9ccc65;">"challenge1.ciphertext.txt"</span>, quote=F)
</pre>
</div>

<p>
To run a GWA use the <a href="https://github.com/encryption4genetics/mixed-model-gwas">mixed-model-gwas</a> and <a href="https://github.com/encryption4genetics/hegp-R">hegp-R</a> packages you can run in R
</p>

<div class="org-src-container">
<pre class="src src-r"><span style="color: #8bc34a;">library</span>(hegp)
<span style="color: #8bc34a;">load</span>(<span style="color: #9ccc65;">"challenge2.ciphertext.RData"</span>)
ge.mm=basic.mm.gwas(De,mc.cores=1)
  Built kinship dim 1586 1586
  estimated heritability 4.355436e-05
ls(ge.mm)
  [1] <span style="color: #9ccc65;">"bp"</span>   <span style="color: #9ccc65;">"chr"</span>  <span style="color: #9ccc65;">"logP"</span>
ge.mm$logP[0:6]
  [1] 0.22838514 0.74881022 0.13431556 0.00929719 1.13027222 0.13211178
</pre>
</div>

<p>
and draw a plot with
</p>

<div class="org-src-container">
<pre class="src src-r"><span style="color: #8bc34a;">library</span>(ggplot2)
ggplot(ge.mm) + geom_point(aes(bp, logP), colour=<span style="color: #9ccc65;">"#000099"</span>)
</pre>
</div>

<p>
 <img width="100%" src="challenge2.ge.mm.png" />
</p>

<p>
To validate we ran the encrypted version (De) and then plaintext
version (D) to compare outcomes with
</p>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #8bc34a;">library</span>(hegp)
<span style="color: #8bc34a;">load</span>(<span style="color: #9ccc65;">"challenge2.ciphertext.RData"</span>)
<span style="color: #8bc34a;">load</span>(<span style="color: #9ccc65;">"challenge2.plaintext.RData"</span>)
ls()
  [1] <span style="color: #9ccc65;">"D"</span>  <span style="color: #9ccc65;">"De"</span>
ls(D)
  [1] <span style="color: #9ccc65;">"cov"</span>  <span style="color: #9ccc65;">"geno"</span> <span style="color: #9ccc65;">"maf"</span>  <span style="color: #9ccc65;">"map"</span>  <span style="color: #9ccc65;">"y"</span>
ls(De)
  [1] <span style="color: #9ccc65;">"cov"</span>  <span style="color: #9ccc65;">"geno"</span> <span style="color: #9ccc65;">"map"</span>  <span style="color: #9ccc65;">"y"</span>
g=basic.gwas(D)
ge=basic.gwas(De)
mean(abs(g$logP-ge$logP))
  0.02951522
ge.mm=basic.mm.gwas(De,mc.cores=1)
  Built kinship dim 1586 1586
  estimated heritability 4.355436e-05
g.mm=basic.mm.gwas(D,mc.cores=1)
  Built kinship dim 1586 1586
  estimated heritability 4.356434e-05
mean(abs(g.mm$logP-ge.mm$logP))
   [1] 0.02943622
</pre>
</div>

<p>
Documentation for the R encryption and GWA functions can be found
<a href="https://github.com/encryption4genetics/hegp-R/blob/master/hegp.pdf">here</a>.
</p>
</div>
</div>


<div id="outline-container-orgd1fd6d2" class="outline-2">
<h2 id="orgd1fd6d2"><span class="section-number-2">3.</span> Reference code</h2>
<div class="outline-text-2" id="text-3">
<p>
The reference code for HEGP is published under <a href="https://www.gnu.org/licenses/gpl-3.0.en.html">GPLv3</a> licensed <a href="https://github.com/encryption4genetics/hegp-R">R code</a>
and <a href="https://github.com/encryption4genetics/hegp-julia">Julia code</a>. An example of running a GWA as presented in the paper
can be found <a href="https://github.com/encryption4genetics/HEGP">here</a>. The algorithm with a description of brute force
attack is described in the <a href="https://www.genetics.org/content/215/2/359#sec-4">results section</a>.
</p>
</div>
</div>

<div id="outline-container-orgfa78c7c" class="outline-2">
<h2 id="orgfa78c7c"><span class="section-number-2">4.</span> Submissions</h2>
<div class="outline-text-2" id="text-4">
<p>
Submissions should be posted in a permanent public git account
(GitHub, gitlab or similar) and reward will be given in USD.  The
solution should be reproducible and announced on the website <a href="https://github.com/encryption4genetics/HEGP-website/issues">issue
tracker</a>. In case you don't want to use the issue tracker it is also
possible to E-mail the authors to indicate where the solution is
hosted. One such E-mail address is pjotr.public542 at thebird dot nl.
</p>

<p>
Important note: only real solutions will be considered that can
reliably crack the key of any such dataset. If the original key
somehow gets ripped from one of our laptops, for example, it does not
count. Security stands or falls with how data is handled and how a key
is generated. Even so, this challenge is about algorithmically
cracking the code for the general case.
</p>
</div>
</div>

<div id="outline-container-orga65d5e4" class="outline-2">
<h2 id="orga65d5e4"><span class="section-number-2">5.</span> Feedback</h2>
<div class="outline-text-2" id="text-5">
<p>
Feedback on the challenge, this website etc., can be posted to the
<a href="https://github.com/encryption4genetics/HEGP-website/issues">issue tracker</a>.
</p>
</div>
</div>

<div id="outline-container-org24f8220" class="outline-2">
<h2 id="org24f8220"><span class="section-number-2">6.</span> General introduction</h2>
<div class="outline-text-2" id="text-6">
<p>
The homomorphic encryption method consists of an orthogonal
transformation by multiplication by the orthogonal matrix P with a
matrix containing the data y and G.
</p>

<p>
 <img src="slide.jpg" width="800" />
</p>

<p>
From the original <a href="https://www.genetics.org/content/215/2/359#sec-4">paper</a>: privacy in relation to quantitative genetic
analysis. (A) A numeric phenotype vector y (left) and genotype dosage
matrix G (right) are represented as colours and shades of gray. Each
row of the matrix represents one individual and each column one
SNP. Genotypes are encoded as imputed dosages clustered at the values
Embedded Image giving the numbers of alternate alleles. (B) The same
data after multiplication by an orthogonal matrix P (a rotation
represented by the curved orange arrow). The genotype dosages are now
represented by a continuum of real numbers. (C) The distribution of
dosages for a particular SNP (column of G), clustered around
0,1,2. (D) The distribution of the same dosages after orthogonal
transformation by multiplication by the orthogonal matrix P (black
histogram) with the normal distribution with same mean and variance
superimposed in red. (F) The normal QQ plot for the data in D, showing
the transformed dosages are very close to a normal distribution. (E) A
cartoon of the HEGP scheme. The top black arrow and equation show the
linear mixed model relating the phenotype y to genotype G with
regression coefficients β representing the allelic effects. The
variance matrix for the residuals is V. After multiplication by
orthogonal matrix P, plaintext data y, G  and the
mixed linear model are transformed as shown in orange. The likelihood
and regression estimates  β are preserved. HEGP,
homomorphic encryption for genotypes and phenotypes; QQ,
quantile–quantile.
</p>
</div>
</div>

<div id="outline-container-orgf59437b" class="outline-2">
<h2 id="orgf59437b"><span class="section-number-2">7.</span> Toy Example</h2>
<div class="outline-text-2" id="text-7">
<p>
The task is best described with an small example, a SNP dosage matrix G with 10 subjects (rows) and 8 SNPs (columns):
</p>

<pre class="example">
sub1     0    1    2    1    1    1    1    2
sub2     0    1    2    1    2    2    2    2
sub3     1    1    2    1    2    2    2    1
sub4     2    1    2    2    1    2    1    1
sub5     2    2    2    0    1    1    2    1
sub6     2    2    2    1    1    2    2    2
sub7     1    2    1    1    1    1    2    2
sub8     0    1    2    1    2    1    2    2
sub9     1    1    2    2    1    1    1    2
sub10    1    2    2    1    2    2    2    1
</pre>

<p>
The first step standardises each column to have mean 0 and variance 1. This does not encrypt the data but it makes the encryption easier to process. The resulting plaintext dosage matrix H is
</p>

<pre class="example">
sub1  -1.224745e+00 -0.7745967  0.3162278 -0.1761661 -0.7745967 -0.9486833 -1.449138  0.7745967
sub2  -1.224745e+00 -0.7745967  0.3162278 -0.1761661  1.1618950  0.9486833  0.621059  0.7745967
sub3  -5.749536e-10 -0.7745967  0.3162278 -0.1761661  1.1618950  0.9486833  0.621059 -1.1618950
sub4   1.224745e+00 -0.7745967  0.3162278  1.5854946 -0.7745967  0.9486833 -1.449138 -1.1618950
sub5   1.224745e+00  1.1618950  0.3162278 -1.9378267 -0.7745967 -0.9486833  0.621059 -1.1618950
sub6   1.224745e+00  1.1618950  0.3162278 -0.1761661 -0.7745967  0.9486833  0.621059  0.7745967
sub7  -1.761818e-09  1.1618950 -2.8460499 -0.1761661 -0.7745967 -0.9486833  0.621059  0.7745967
sub8  -1.224745e+00 -0.7745967  0.3162278 -0.1761661  1.1618950 -0.9486833  0.621059  0.7745967
sub9   5.584054e-10 -0.7745967  0.3162278  1.5854946 -0.7745967 -0.9486833 -1.449138  0.7745967
sub10 -7.763129e-10  1.1618950  0.3162278 -0.1761661  1.1618950  0.9486833  0.621059 -1.1618950
</pre>

<p>
Next, we sample a random orthogonal 10x10 matrix P (the key)
</p>

<pre class="example">
 [1,] -0.02512827 -0.4797328  0.07364427  0.29302653 -0.52531836 -0.06077012 -0.035656119 -0.21264789 -0.03656364  0.59199429
 [2,]  0.49690414 -0.2130092 -0.18437213 -0.08094409 -0.33989098  0.11458794  0.382316649  0.51918534 -0.29625914 -0.18714471
 [3,]  0.23650365  0.2027581  0.09184543  0.01711933  0.45947279 -0.11898452  0.549979207  0.04901441  0.03087585  0.60259531
 [4,] -0.12668661  0.2640016 -0.24657807 -0.31047136 -0.08100156  0.01898761 -0.009330209 -0.34743514 -0.77909552  0.14950124
 [5,]  0.55558644 -0.1256327 -0.11075554 -0.35350803  0.12981113  0.51028646 -0.332763562 -0.28269321  0.20446208  0.16914618
 [6,] -0.20079972 -0.1310010  0.33769938 -0.45070382  0.09236098 -0.05672098 -0.390626201  0.58184807 -0.14355153  0.31914343
 [7,]  0.30606137  0.1956211  0.67114299 -0.30536330 -0.32325985 -0.31468499  0.106176504 -0.27993855  0.04238805 -0.17152201
 [8,] -0.31057809 -0.5813563  0.28456863 -0.21173902  0.26275032  0.30931163  0.411201611 -0.21696952 -0.12053771 -0.21059453
 [9,]  0.07808339 -0.3843085 -0.42321573 -0.37251969  0.13692025 -0.68291915 -0.007551188 -0.12694060  0.15450958 -0.05619238
[10,] -0.36775701  0.2383173 -0.23348592 -0.45381168 -0.41614295  0.20969099  0.325976664  0.04578156  0.44759290  0.14716487
</pre>

<p>
Then we encrypt the dosages F = PH to make the ciphertext
</p>

<pre class="example">
sub1   0.5198393  0.26551339 -0.01916052  1.4507213  0.2713387  1.18915219 -0.7379842 -1.16229696
sub2  -1.3586430 -0.22207809 -1.14214913 -0.1029684 -0.2902612 -1.24575229 -0.1165288  1.69811232
sub3  -0.1600357  1.24824503 -1.06841109 -1.0985638  0.1892755 -0.50328813  0.7284091 -0.62463676
sub4  -0.1988559  1.28854213 -0.43475181 -1.5181190  0.7876347  1.15643550  1.6061022 -0.19111906
sub5   0.1706426  0.64079385  1.16737846 -0.5553653 -0.9595529 -0.17537980 -0.6155911  0.60202029
sub6  -0.8145915 -0.03660399  1.22187582 -1.2021241  2.1778375  0.07512188  1.6196177 -0.61084795
sub7  -1.4268916 -1.30507703 -0.35896396  0.1191278  0.8610707  0.21228333 -0.1347692  0.19297276
sub8   1.7994281  1.79366813 -1.42174710 -0.9805980 -1.1053001 -0.41331874  1.0923847 -0.53943417
sub9  -0.5944334  0.12376342 -0.50869100 -0.3285796 -0.6138749 -2.04363166 -0.7562708  0.08008153
sub10 -0.7061947  0.56034104 -1.04875037  0.7321312  0.4268941 -0.12102879  0.7390060  1.80792099
</pre>

<p>
This form of encryption is homomorphic with respect to many quantitative genetics analyses, particularly the mixed model GWAS.
</p>

<p>
To decrypt the ciphertext it is necessary to multiply it by the inverse of the key P, which is equal to the transpose of P because it is orthogonal.
</p>

<p>
The challenge is, in the absence of knowing either P or H, to find an
orthogonal matrix Q such that QF "looks like" a genotype dosage
matrix. That is, the distribution of the plaintext for a given SNP
will be trimodal (or bimodal is the rarer homozygote genotype is
absent from the sample) with expected modes specified by the
Hardy-Weinberg equilibrium distribution. We provide the allele
frequencies of the plaintext, which will help in defining these
expected modes (see the maf.txt files). It is reasonable to
provide this information even though it might make the encryption less
secure because users of the ciphertext would need this information for
some genetic analyses.
</p>

<p>
We provide the toy dataset and its encryption key as R objects in the
file "toy.hegp.RData" from <a href="https://files.genenetwork.org/Toy.HEGP/">Toy-HEGP</a> so that you can reproduce these
analyses. E.g.
</p>

<div class="org-src-container">
<pre class="src src-R"><span style="color: #b0bec5;"># </span><span style="color: #b0bec5;">Load a table</span>
toy = read.table(<span style="color: #9ccc65;">"toy2.txt"</span>,header=F,row.names=1)

<span style="color: #8bc34a;">library</span>(hegp)
<span style="color: #8bc34a;">load</span>(<span style="color: #9ccc65;">"toy.hegp.RData"</span>) <span style="color: #b0bec5;"># </span><span style="color: #b0bec5;">loads encryption key</span>
plaintext.toy$geno = as.matrix(toy)
c = encrypt.D(plaintext.toy, key.toy)
d = encrypt.D(c, key.toy, invert=<span style="color: #84ffff;">TRUE</span>)
mean(abs(c$geno-d$geno))
[1] 2.524352e-09
</pre>
</div>
</div>
</div>

<div id="outline-container-org2c4b78f" class="outline-2">
<h2 id="org2c4b78f"><span class="section-number-2">8.</span> References</h2>
<div class="outline-text-2" id="text-8">
<ul class="org-ul">
<li>The original HEGP <a href="https://www.genetics.org/content/215/2/359">paper</a></li>
<li>HEGP discussion on <a href="https://redd.it/jn8g9h">reddit crypto forum</a></li>
<li>For more background see this <a href="./start">page</a></li>
</ul>
</div>
</div>
</div>
<div id="postamble" class="status">
<hr><small>Created by <a href="http://thebird.nl/">Pjotr Prins</a> (pjotr.public768 at thebird 'dot' nl) using Emacs org-mode and a healthy dose of Lisp!<br />Modified 2024-03-30 Sat 03:25</small>.
</div>
</body>
</html>