<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2020-10-06 Tue 11:03 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>&lrm;</title>
<meta name="generator" content="Org mode" />
<meta name="author" content="Pjotr Prins" />
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .equation-container {
    display: table;
    text-align: center;
    width: 100%;
  }
  .equation {
    vertical-align: middle;
  }
  .equation-label {
    display: table-cell;
    text-align: right;
    vertical-align: middle;
  }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2020 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<div id="outline-container-org438bd9e" class="outline-2">
<h2 id="org438bd9e">The $1,000 Challenge!</h2>
<div class="outline-text-2" id="text-org438bd9e">
<p>
<i>To go straight to the actual challenge click
<a href="./challenge">here</a></i>.
</p>

<p>
Crack the code! The authors announce the HEGP (heegeepee;
<i>hiːdʒiːpiː</i>) Challenge,  <img width="80" src="stethoscope.png"
align="right" /> with a $1,000 (one thousand dollar) prize for
the individual or group who can crack HEGP encrypted data. HEGP, when
proven solid, will have a large impact on the way human genetics is
pursued today because it will allow for sharing genotype data while
preserving privacy of the individual. Not only is sharing data
required for reproducible Science, there is also a large interest in
hosting data on laptops and servers that need not be <a href="https://en.wikipedia.org/wiki/Health_Insurance_Portability_and_Accountability_Act">HIPAA</a> compliant.
</p>

<p>
On this page we chat about cracking HEGP, Rubik cube, DNA,
strawberries and Enigma.
</p>

<p>
<i>To go straight to the actual challenge click
<a href="./challenge">here</a></i>.
</p>
</div>
</div>

<div id="outline-container-org04c12dc" class="outline-2">
<h2 id="org04c12dc">Relevance</h2>
<div class="outline-text-2" id="text-org04c12dc">
<p>
With a <a href="https://www.genetics.org/content/215/2/359">publication in Genetics</a> we present a novel Homomorphic
encryption method named HEGP that allows for sharing of genotypes and
phenotypes in the context of Genome Wide Association (GWA)
studies. Importantly sharing but without giving away <b>private</b>
information and making individuals identifiable. For more information
check out <a href="https://www.slideshare.net/EmilianoDC/the-genomics-revolution-the-good-the-bad-and-the-ugly">a privacy researcher's perspective</a>.
</p>

<p>
With HEGP anyone can freely hand over genotype and phenotype data to
anyone else and have them reproduce the results. Even better, the
encoding can be stacked. So the second party can encrypt and combine
their data independently and give it to a third party.
</p>

<p>
So any party can use our data to add it their own analysis. And they
can share their data back with us. As such, genotype data can be
stored on (public) servers and GWA analysis can be reproduced.
</p>

<p>
This is a breakthrough in <a href="https://en.wikipedia.org/wiki/FAIR_data">FAIR data</a> sharing and contrasts greatly with
current practice of hiding/protecting genotype data and only providing
summary statistics. One example is an important UK Biobank <a href="https://datashare.is.ed.ac.uk/handle/10283/3083">depression</a>
study where genetic markers are presented with their statistical
significance. We can only assume that this study can be reproduced by
a group of researchers having access to the original data. The truth
is that these outcomes can not be reproduced by you or me! HEGP will
make data sharing and reproducible analysis a <b>reality</b>.
</p>
</div>
</div>

<div id="outline-container-org69451a6" class="outline-2">
<h2 id="org69451a6">Simple explanation</h2>
<div class="outline-text-2" id="text-org69451a6">
<p>
Homomorphic encryption is a mathematical translation of data into an
encrypted form where the result of a computation is the same for the
unencrypted and encrypted forms. With HEGP a matrix of data is
transformed by a high-dimensional random linear orthogonal
transformation key as described in the <a href="https://www.genetics.org/content/215/2/359">paper</a> (open access) and
visualised in below animation (hit  <span class="buttons">
<button id="playtxt">Play</button> </span>). The resulting matrix
scrambles the data while preserving the 'shape of the data' for
analysis.   <img width="80" src="rubik.jpg" align="right" />
One way to think about this is that when a Rubik cube gets rotated the
fields change colour, but the object still maintains its shape as a
three dimensional cube.  With DNA the genotype/phenotype shape is
typically used to predict associations between genotypes and
phenotypes.   <img width="80" src="strawberry.jpg"
align="right" /> An example of a phenotype is a preference for
strawberry taste. An example of an associated genotype is a DNA
encoded <a href="https://en.wikipedia.org/wiki/Olfactory_receptor">olfactory receptor</a>.
</p>

<p>
Here we display a typical example of genome-wide association (GWA) of
phenotype against genotype:
</p>

<p>
 <img src="gemma.gif" width="100%" alt="Example of genome-wide
association (GWA) of phenotype against genotype" />
</p>

<p>
Genes (on chromosomes) involved in some trait are marked. This is the
backbone computation for finding genes involved in some trait and
pursued in the <a href="https://www.ukbiobank.ac.uk/">UK Biobank</a> involving half a million subjects. To find
associations GWA is applied to find genes involved in, for example,
cancer or COVID-19 mortality; i.e., the first step towards finding
causality and potentially better treatments.
</p>

<p>
 <img src="https://www.genetics.org/content/genetics/215/2/359/F1.large.jpg" width="100%" />
</p>

<p>
In above image data is shown before and after encryption. The
unencrypted data contains three values while the encrypted data shows
a normal distribution.
</p>
</div>
</div>

<div id="outline-container-org1d43cda" class="outline-2">
<h2 id="org1d43cda">Enigma and why a challenge?</h2>
<div class="outline-text-2" id="text-org1d43cda">
<p>
The Enigma machine is an encryption device developed and used in the
20th century to protect commercial, diplomatic and military
communication. It was employed extensively by Nazi Germany during
World War II, in all branches of the German military (source
<a href="https://en.wikipedia.org/wiki/Enigma_machine">wikipedia</a>).   <img src="enigma.jpg" align="left" /> Enigma
encrypted text by a transformation and was cracked by the Polish
Cipher Bureau in 1932 and the crack was used by the allied forces to
win the war.  To ascertain HEGP is bullet proof, unlike ENIGMA, we
invite the algorithmic inclined to crack the code and make HEGP
history (one way or the other).
</p>
</div>
</div>
</div>
<div id="postamble" class="status">
<hr><small>Created by <a href="http://thebird.nl/">Pjotr Prins</a> (pjotr.public768 at thebird 'dot' nl) using Emacs org-mode and a healthy dose of Lisp!<br />Modified 2020-10-06 Tue 11:02</small>.
</div>
</body>
</html>
